src/golioth_fw_block_processor.c

```diff
@@ -155,7 +159,41 @@ static golioth_status_t decompress(const uint8_t* in_data, size_t in_data_size,
         } while (pres == HSDR_POLL_MORE);
     }
 #elif CONFIG_GOLIOTH_OTA_ZLIB_DECOMPRESSION_ENABLE == 1
-    // TODO - miniz_tinfl()
+    GLTH_LOGD(TAG, "tinfl_decompress input: %zd", in_data_size);
+
+    size_t in_buf_ofs = 0;
+    tinfl_status status = TINFL_STATUS_HAS_MORE_OUTPUT;
+    while (status == TINFL_STATUS_HAS_MORE_OUTPUT) {
+        // This is basically the inner loop of tinfl_decompress_mem_to_callback() in miniz_tinfl.c
+        size_t in_buf_size = in_data_size - in_buf_ofs;
+        size_t dst_buf_size = TINFL_LZ_DICT_SIZE - ctx->dict_ofs;
+        status = tinfl_decompress(
+                &ctx->decompressor,
+                (const mz_uint8*)in_data + in_buf_ofs,
+                &in_buf_size,
+                _zlib_dict,
+                _zlib_dict + ctx->dict_ofs,
+                &dst_buf_size,
+                TINFL_FLAG_PARSE_ZLIB_HEADER
+                        | (in_data_size == 1024 ? TINFL_FLAG_HAS_MORE_INPUT : 0));
```

Comment:
How about spliting that in to separate:
```
mz_uint32 flags = TINFL_FLAG_PARSE_ZLIB_HEADER;

if (in_data_size == 1024) {
	flags |= TINFL_FLAG_HAS_MORE_INPUT;
}
```
as it is kind of busy in the invocation of `tinfl_decompress()`.

I think we should somehow improve `in_data_size == 1024` check. This is very arbitrary, very CoAP specific and very CoAP implementation specific (that the block size will be 1024 bytes). I think so, but maybe there is something more that I am missing?

Reply:
:+1: to `is_last` flag. Alternatively there could be separate functions called on the beginning, when processing and at the end, something like:
- `*_begin`
- `*_process`
- `*_end`

Whichever works better in this context.

Reply:
In testing, it seems like unconditionally setting the flags to this works just fine:

```c
mz_uint32 flags = TINFL_FLAG_PARSE_ZLIB_HEADER | TINFL_FLAG_HAS_MORE_INPUT
```

By doing this, we are telling the decompressor there is always more input to come. This is a lie on the last block, but it doesn't seem to be a problem for valid compressed artifacts, and the full artifact is properly decompressed.

I've tested some malformed compressed artifacts, and it properly reports an error, so I don't think setting the flags this way would inhibit our ability to detect errors.

So I think I'd like to go with this solution. The `is_last` and `*_{begin,process,end}` are both fairly complicated to do, and I want to avoid it unless it's strictly necessary.